sky/skylet/providers/gcp/config.py

```diff
@@ -504,24 +504,45 @@ def _check_firewall_rules(vpc_name, config, compute):
         return False
     effective_rules = response["firewalls"]
 
-    def _get_refined_rule(rule):
-        KEY_TO_COMPARE = {"sourceRanges", "allowed", "direction"}
-        refined_rule = {}
-        for k in KEY_TO_COMPARE:
-            if k not in rule:
-                continue
-            if k == "allowed":
-                refined_rule[k] = sorted(rule[k], key=lambda x: x["IPProtocol"])
-            else:
-                refined_rule[k] = rule[k]
-        return refined_rule
-
-    required_rules = list(map(_get_refined_rule, required_rules))
-    effective_rules = list(map(_get_refined_rule, effective_rules))
-
-    for rule in required_rules:
-        if rule not in effective_rules:
+    def _merge_and_refine_rule(rules):
+        # Output: dict of (direction, source) -> {protocol -> ports}
+        tmp = {}
+        for rule in rules:
+            direction = rule.get("direction", "")
+            sources = rule.get("sourceRanges", [])
+            allowed = rule.get("allowed", [])
+            for source in sources:
+                tmp[(direction, source)] = tmp.get((direction, source), []) + allowed
+        source2rules = {}
+        for (direction, source), allowed_list in tmp.items():
+            source2rules[(direction, source)] = {}
+            for allowed in allowed_list:
+                ports = set()
+                for port in allowed.get('ports', set()):
+                    parse_ports = port.split('-')
+                    if len(parse_ports) == 1:
+                        ports.add(int(parse_ports[0]))
+                    else:
+                        ports.update(
```

Comment:
```suggestion
                        assert len(parse_ports) == 2, parse_ports
                        ports.update(
```

Reply:
Ah thanks! I also fixed a case when port list is empty. which indeed means all ports are open..
https://github.com/skypilot-org/skypilot/pull/1640/files#diff-48dd58bcac2e32b383ee0b1d09abfe22fbad89e1666ee7369f4e6dbc7102ff84R527